3.13/redundancy_eliminator.md

```diff
@@ -0,0 +1,126 @@
+# The Tier 2 Redundancy Eliminator
+
+Author: Ken Jin, Mark Shannon
+
+Tier 2 traces require optimization. At minimum, redundant operations
+should be removed. This is where the name "Redundancy Eliminator" is
+derived from. This forms part of the larger effort towards
+the tier 2 optimizer.
+
+## Abstract Interpretation of Uops
+
+To analyze the uops, [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation)
+shall be used.
+
+Just like the CPython interpreter is an interpretation over
+Python program state, represented by `PyObject *`, the abstract interpreter
+shall be an interpretation over Python symbolic information state,
+represented by `_Py_UOpsSymType *`.
+
+The abstract interepeter has the equivalent constructs of the
+CPython interpreter to represent state. It has symbolic
+local and symbolic frames, This allows it to emulate CPython
+runtime state accurately.
+
+Executing "steps" in the abstract interpreter is thus the same as
+doing the same in CPython -- by interpreting bytecode instructions.
+The abstract interpreter thus executes a bytecode, update its
+symbolic state, and repeats until it reaches a terminator (the end
+of the trace).
+
+### Lattice
+
+**Note: if you are uninterested in the theory behind data-flow analysis,\
+please skip this section.**
+
+The states of information in such an abstract interpretation can be
+represented by a [lattice](https://en.wikipedia.org/wiki/Lattice_(order))
+of values. The lattice itself is an ordered powerset
+of all variables and their associated information.
+
+For example, here is a trivial lattice for the statement 
+```python
+x = 1
+y = 2.0
+```
+:
+
+```mermaid
+flowchart TD
+    A("`Bottom, ie. x: *?*, y: *?*`") --> B("`x: *int*, y: *?*`")
+    A --> C("`x: *?*, y: *float*`")
+    B --> D("`Top, x: *int*, y: *float*`")
+    C --> D
+  
+```
+
+Real Python runtime information
+may contain more complex types, constants, and unions of types.
+
+Each `_Py_UOpsSymType` captures all the information associated
+with a variable at any point in the lattice.
```

Comment:
It would be nice to go over at least the most basic info contained in this struct (the type and the constant -- you can leave out refinements and the fact that the type may be a union).